PSC register failure: com.vmware.vim.vmomi.client.common.UnexpectedStatusCodeException: Unexpected status code: 400

This is a copy of an issue [196](https://github.com/vmware/admiral/issues/196) initially opened in Admiral product:

----

Found by a customer in OVA RC3 build https://storage.googleapis.com/vic-product-ova-builds/vic-v1.2.0-rc3-ce9fca67.ova

Upon running `admiral-auth-psc-1.2.0-SNAPSHOT-command.jar --command=register --tenant=<non " .local" domain>  ... (other args)`, it fails with:

```
[0][I][2017-09-07T19:53:20.378Z][1][SsoPscCommand][register][Initializing...]
[1][I][2017-09-07T19:53:20.385Z][1][SsoPscCommand][register][Registering...]
[2][I][2017-09-07T19:53:20.409Z][1][SsoPscCommand][deleteFile][Removal of file: '/etc/vmware/psc/harbor/psc-config.keystore' result: true]
[3][I][2017-09-07T19:53:31.070Z][1][VmodlContextImpl$NonValidatingClassPathXmlApplicationContext][prepareRefresh][Refreshing com.vmware.vim.vmomi.core.types.impl.VmodlContextImpl$NonValidatingClassPathXmlApplicationContext@59e5ddf: startup date [Thu Sep 07 19:53:31 UTC 2017]; root of context hierarchy]
[4][I][2017-09-07T19:53:31.097Z][1][XmlBeanDefinitionReader][loadBeanDefinitions][Loading XML bean definitions from class path resource [com/vmware/vim/binding/vmodl/context_v2.xml]]
[5][I][2017-09-07T19:53:31.294Z][1][VmodlContextImpl$NonValidatingClassPathXmlApplicationContext][doClose][Closing com.vmware.vim.vmomi.core.types.impl.VmodlContextImpl$NonValidatingClassPathXmlApplicationContext@59e5ddf: startup date [Thu Sep 07 19:53:31 UTC 2017]; root of context hierarchy]
[6][I][2017-09-07T19:53:31.296Z][1][VmodlContextImpl][loadVmodlPackage][Package com.vmware.vim.binding.vmodl loaded in 243 millis]
[7][I][2017-09-07T19:53:31.296Z][1][VmodlContextImpl$NonValidatingClassPathXmlApplicationContext][prepareRefresh][Refreshing com.vmware.vim.vmomi.core.types.impl.VmodlContextImpl$NonValidatingClassPathXmlApplicationContext@6e509ffa: startup date [Thu Sep 07 19:53:31 UTC 2017]; root of context hierarchy]
[8][I][2017-09-07T19:53:31.297Z][1][XmlBeanDefinitionReader][loadBeanDefinitions][Loading XML bean definitions from class path resource [com/vmware/vim/binding/vmodl/context_v2.xml]]
[9][I][2017-09-07T19:53:31.324Z][1][VmodlContextImpl$NonValidatingClassPathXmlApplicationContext][doClose][Closing com.vmware.vim.vmomi.core.types.impl.VmodlContextImpl$NonValidatingClassPathXmlApplicationContext@6e509ffa: startup date [Thu Sep 07 19:53:31 UTC 2017]; root of context hierarchy]
[10][I][2017-09-07T19:53:31.324Z][1][VmodlContextImpl][loadVmodlPackage][Package com.vmware.vim.binding.vmodl loaded in 27 millis]
[11][I][2017-09-07T19:53:31.325Z][1][VmodlContextImpl$NonValidatingClassPathXmlApplicationContext][prepareRefresh][Refreshing com.vmware.vim.vmomi.core.types.impl.VmodlContextImpl$NonValidatingClassPathXmlApplicationContext@6236eb5f: startup date [Thu Sep 07 19:53:31 UTC 2017]; root of context hierarchy]
[12][I][2017-09-07T19:53:31.325Z][1][XmlBeanDefinitionReader][loadBeanDefinitions][Loading XML bean definitions from class path resource [com/vmware/vim/binding/sso/context.xml]]
[13][I][2017-09-07T19:53:31.484Z][1][VmodlContextImpl$NonValidatingClassPathXmlApplicationContext][doClose][Closing com.vmware.vim.vmomi.core.types.impl.VmodlContextImpl$NonValidatingClassPathXmlApplicationContext@6236eb5f: startup date [Thu Sep 07 19:53:31 UTC 2017]; root of context hierarchy]
[14][I][2017-09-07T19:53:31.484Z][1][VmodlContextImpl][loadVmodlPackage][Package com.vmware.vim.binding.sso loaded in 159 millis]
[15][W][2017-09-07T19:53:31.559Z][1][SiteAffinityServerEndpointProvider][logWarning][CDC not configured java.lang.NoClassDefFoundError: com/vmware/identity/cdc/CdcFactory]

Exception in thread "main" com.vmware.admiral.auth.idm.psc.saml.sso.admin.SsoAdminClient
Exception: General SSO failure. Invalid host, port or tenant.
at com.vmware.admiral.auth.idm.psc.saml.sso.admin.SsoAdminClientFactoryImpl.createSSOAdminClient(SsoAdminClientFactoryImpl.java:148)
at com.vmware.admiral.auth.idm.psc.saml.sso.admin.SsoAdminFacade.registerOrRetrieveSolutionUser(SsoAdminFacade.java:172)
at com.vmware.admiral.auth.idm.psc.saml.util.SsoPscCommand.register(SsoPscCommand.java:83)
at com.vmware.admiral.auth.idm.psc.util.PscCommand.execute(PscCommand.java:37)
at com.vmware.admiral.auth.idm.psc.util.PscCommand.main(PscCommand.java:47)
Caused by: com.vmware.vim.sso.admin.exception.InternalError: General failure.
at com.vmware.vim.sso.admin.client.vmomi.impl.VmomiClientCommand.execute(VmomiClientCommand.java:211)

at com.vmware.vim.sso.admin.client.vmomi.impl.VmomiClientCommand.executeEnsuringNoDomainError(VmomiClientCommand.java:217)
at com.vmware.vim.sso.admin.client.vmomi.impl.AdminClientImpl.createServiceContent(AdminClientImpl.java:334)
at com.vmware.vim.sso.admin.client.vmomi.impl.AdminClientImpl.<init>(AdminClientImpl.java:107)
at com.vmware.vim.sso.admin.client.vmomi.VmomiClientFactory.createAdminClient(VmomiClientFactory.java:64)
at com.vmware.vim.sso.admin.client.vmomi.VmomiClientFactory.createAdminClient(VmomiClientFactory.java:54)
at com.vmware.admiral.auth.idm.psc.saml.sso.admin.SsoAdminClientFactoryImpl.createSSOAdminClient(SsoAdminClientFactoryImpl.java:123)
... 4 more

Caused by: com.vmware.vim.vmomi.client.common.UnexpectedStatusCodeException: Unexpected status code: 400
at com.vmware.vim.vmomi.client.common.Response$Status.getStatus(Response.java:58)
at com.vmware.vim.vmomi.client.http.impl.HttpExchangeBase.parseResponse(HttpExchangeBase.java:150)
at com.vmware.vim.vmomi.client.http.impl.HttpExchange.run(HttpExchange.java:48)
at com.vmware.vim.vmomi.client.http.impl.HttpProtocolBindingBase.executeRunnable(HttpProtocolBindingBase.java:226)
at com.vmware.vim.vmomi.client.http.impl.HttpProtocolBindingImpl.send(HttpProtocolBindingImpl.java:109)
at com.vmware.vim.vmomi.client.common.impl.MethodInvocationHandlerImpl$CallExecutor.sendCall(MethodInvocationHandlerImpl.java:581)
at com.vmware.vim.vmomi.client.common.impl.MethodInvocationHandlerImpl$CallExecutor.executeCall(MethodInvocationHandlerImpl.java:562)
at com.vmware.vim.vmomi.client.common.impl.MethodInvocationHandlerImpl.completeCall(MethodInvocationHandlerImpl.java:344)
at com.vmware.vim.vmomi.client.common.impl.MethodInvocationHandlerImpl.invokeOperation(MethodInvocationHandlerImpl.java:304)
at com.vmware.vim.vmomi.client.common.impl.MethodInvocationHandlerImpl.invoke(MethodInvocationHandlerImpl.java:178)
at com.sun.proxy.$Proxy25.retrieveServiceContent(Unknown Source)
at com.vmware.vim.sso.admin.client.vmomi.impl.AdminClientImpl$1.actionCommand(AdminClientImpl.java:339)
at com.vmware.vim.sso.admin.client.vmomi.impl.AdminClientImpl$1.actionCommand(AdminClientImpl.java:334)
at com.vmware.vim.sso.admin.client.vmomi.impl.VmomiClientCommand.execute(VmomiClientCommand.java:103)
... 10 more"
``` 

**Analysis:**

This is similar to #147, however, the stack trace shows HTTP code 400, not 404. Also, the user domain exists - upon hitting `https://{psc-hostname}/sso-adminserver/sdk/{tenant-id}`, the response is:

```
<?xml version="1.0" encoding="UTF-8" ?> 
- <soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:soapenc="http://schemas.xmlsoap.org/soap/encoding/" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
- <soapenv:Body>
- <soapenv:Fault>
  <faultcode>ServerFaultCode</faultcode>
  <faultstring>Unexpected EOF in prolog at [row,col {unknown-source}]: [1,0]</faultstring> 
- <detail>
  <RuntimeFaultFault xsi:type="vim25:InvalidRequest" xmlns="urn:vim25" xmlns:vim25="urn:vim25" />
  </detail>
  </soapenv:Fault>
  </soapenv:Body>
  </soapenv:Envelope>
```

which according to https://github.com/vmware/admiral/issues/147#issuecomment-320801876 means that this domain exists in PSC (cc @shadjiiski)

The customer has also attempted other user domains, including AD domains. What could be missing here?

@martin-borisov @sergiosagu @lcastellano Please help triage this, thanks!

--- a/installer/fileserver/html/index.html
+++ b/installer/fileserver/html/index.html
@@ -214,6 +214,8 @@
                                         <div class="form-group">
                                             <label for="psc">PSC Instance, if external</label>
                                             <input id="psc" type="text" name="psc" placeholder="FQDN">
+                                            <label for="pscDomain">PSC Admin Domain</label>
+                                            <input id="pscDomain" type="text" name="pscDomain" placeholder="vsphere.local">
                                         </div>
                                     </div>
                                     <div class="text-sm-right">

--- a/installer/fileserver/html/index.html
+++ b/installer/fileserver/html/index.html
@@ -200,7 +200,7 @@
                                         {{ end }}
                                         <p class="mt-0 mb-1">
                                             Please enter your vCenter credentials to complete the vSphere Integrated
-                                            Containers Appliance installation.
+                                            Containers Appliance installation and PSC settings if using an external instance.
                                         </p>
                                         <div class="form-group">
                                             <label for="target" class="required">vCenter Server</label>
@@ -212,11 +212,11 @@
                                             <input type="password" name="password" placeholder="Password" />
                                         </div>
                                         <div class="form-group">
-                                            <label for="psc">PSC Instance, if external</label>
+                                            <label for="psc">External PSC Instance</label>
                                             <input id="psc" type="text" name="psc" placeholder="FQDN">
                                         </div>
                                         <div class="form-group">
-                                            <label for="pscDomain">PSC Admin Domain</label>
+                                            <label for="pscDomain">External PSC Admin Domain</label>
                                             <input id="pscDomain" type="text" name="pscDomain" placeholder="vsphere.local">
                                         </div>
                                     </div>

--- a/installer/fileserver/html/index.html
+++ b/installer/fileserver/html/index.html
@@ -214,6 +214,8 @@
                                         <div class="form-group">
                                             <label for="psc">PSC Instance, if external</label>
                                             <input id="psc" type="text" name="psc" placeholder="FQDN">
+                                        </div>
+                                        <div class="form-group">
                                             <label for="pscDomain">PSC Admin Domain</label>
                                             <input id="pscDomain" type="text" name="pscDomain" placeholder="vsphere.local">
                                         </div>

--- a/installer/fileserver/server.go
+++ b/installer/fileserver/server.go
@@ -260,7 +260,7 @@ func startInitializationServices() string {
 
 	if err := registerWithPSC(ctx); err != nil {
 		log.Debug(errors.ErrorStack(err))
-		errorMsg = append(errorMsg, "Failed to register with PSC. Please check the provided PSC settings and try again")
+		errorMsg = append(errorMsg, "Failed to register with PSC. Please check the PSC settings provided and try again")
 	}
 
 	if len(errorMsg) == 0 {

--- a/installer/fileserver/server.go
+++ b/installer/fileserver/server.go
@@ -255,12 +255,12 @@ func startInitializationServices() string {
 	ctx := context.TODO()
 	if err := tagvm.Run(ctx, admin.Validator.Session); err != nil {
 		log.Debug(errors.ErrorStack(err))
-		errorMsg = append(errorMsg, "Failed to locate productVM, trusted content is not available")
+		errorMsg = append(errorMsg, "Failed to locate VIC Appliance. Please check the vCenter Server provided and try again")
 	}
 
 	if err := registerWithPSC(ctx); err != nil {
 		log.Debug(errors.ErrorStack(err))
-		errorMsg = append(errorMsg, "Failed to register with PSC. Please check the vSphere user domain PSC settings and try again")
+		errorMsg = append(errorMsg, "Failed to register with PSC. Please check the provided PSC settings and try again")
 	}
 
 	if len(errorMsg) == 0 {

--- a/installer/fileserver/server.go
+++ b/installer/fileserver/server.go
@@ -68,6 +68,9 @@ var (
 
 	// pscInstance holds the form input for the PSC field
 	pscInstance string
+
+	// pscDomain holds the form input for the PSC Admin Domain field
+	pscDomain string
 )
 
 const initServicesTimestamp = "./registration-timestamps.txt"
@@ -204,6 +207,7 @@ func indexHandler(resp http.ResponseWriter, req *http.Request) {
 		admin.User = req.FormValue("user")
 		admin.Password = req.FormValue("password")
 		pscInstance = req.FormValue("psc")
+		pscDomain = req.FormValue("pscDomain")
 
 		if err := admin.VerifyLogin(); err != nil {
 			log.Infof("Validation failed: %s", err.Error())

--- a/installer/fileserver/tasks.go
+++ b/installer/fileserver/tasks.go
@@ -56,8 +56,8 @@ func registerWithPSC(ctx context.Context) error {
 			return err
 		}
 	}
-	if pscDomain != "" && domain != pscDomain {
-		log.Warnf("User domain: %s does not match PSC domain: %s. Using %s", domain, pscDomain, pscDomain)
+	if pscDomain != "" {
+		log.Infof("User domain: %s does not match PSC domain: %s. Using %s", domain, pscDomain, pscDomain)
 		domain = pscDomain
 	}
 	log.Infof("vCenter user: %s", admin.User)

--- a/installer/fileserver/tasks.go
+++ b/installer/fileserver/tasks.go
@@ -56,8 +56,8 @@ func registerWithPSC(ctx context.Context) error {
 			return err
 		}
 	}
-	if domain != pscDomain {
-		log.Warnf("User domain/PSC domain mismatch: %s %s", domain, pscDomain)
+	if pscDomain != "" && domain != pscDomain {
+		log.Warnf("User domain: %s does not match PSC domain: %s. Using %s", domain, pscDomain, pscDomain)
 	}
 	log.Infof("vCenter user: %s", admin.User)
 	log.Infof("PSC instance: %s", pscInstance)

--- a/installer/fileserver/tasks.go
+++ b/installer/fileserver/tasks.go
@@ -56,7 +56,12 @@ func registerWithPSC(ctx context.Context) error {
 			return err
 		}
 	}
+	if domain != pscDomain {
+		log.Warnf("User domain/PSC domain mismatch: %s %s", domain, pscDomain)
+	}
+	log.Infof("vCenter user: %s", admin.User)
 	log.Infof("PSC instance: %s", pscInstance)
+	log.Infof("PSC Admin domain: %s", pscDomain)
 
 	// Obtain the OVA VM's IP
 	vmIP, err := ip.FirstIPv4(ip.Eth0Interface)
@@ -95,7 +100,7 @@ func registerWithPSC(ctx context.Context) error {
 			"--clientName=" + client,
 			// NOTE(anchal): version set to 6.0 to use SAML for both versions 6.0 and 6.5
 			"--version=6.0",
-			"--tenant=" + domain,
+			"--tenant=" + pscDomain,
 			"--domainController=" + pscInstance,
 			"--username=" + admin.User,
 			"--password=" + admin.Password,

--- a/installer/fileserver/tasks.go
+++ b/installer/fileserver/tasks.go
@@ -58,6 +58,7 @@ func registerWithPSC(ctx context.Context) error {
 	}
 	if pscDomain != "" && domain != pscDomain {
 		log.Warnf("User domain: %s does not match PSC domain: %s. Using %s", domain, pscDomain, pscDomain)
+		domain = pscDomain
 	}
 	log.Infof("vCenter user: %s", admin.User)
 	log.Infof("PSC instance: %s", pscInstance)
@@ -100,7 +101,7 @@ func registerWithPSC(ctx context.Context) error {
 			"--clientName=" + client,
 			// NOTE(anchal): version set to 6.0 to use SAML for both versions 6.0 and 6.5
 			"--version=6.0",
-			"--tenant=" + pscDomain,
+			"--tenant=" + domain,
 			"--domainController=" + pscInstance,
 			"--username=" + admin.User,
 			"--password=" + admin.Password,

--- a/installer/fileserver/tasks.go
+++ b/installer/fileserver/tasks.go
@@ -57,12 +57,12 @@ func registerWithPSC(ctx context.Context) error {
 		}
 	}
 	if pscDomain != "" {
-		log.Infof("User domain: %s does not match PSC domain: %s. Using %s", domain, pscDomain, pscDomain)
+		log.Infof("User domain: %s PSC domain: %s. Using %s", domain, pscDomain, pscDomain)
 		domain = pscDomain
 	}
 	log.Infof("vCenter user: %s", admin.User)
 	log.Infof("PSC instance: %s", pscInstance)
-	log.Infof("PSC domain: %s", pscDomain)
+	log.Infof("PSC domain: %s", domain)
 
 	// Obtain the OVA VM's IP
 	vmIP, err := ip.FirstIPv4(ip.Eth0Interface)

--- a/installer/fileserver/tasks.go
+++ b/installer/fileserver/tasks.go
@@ -62,7 +62,7 @@ func registerWithPSC(ctx context.Context) error {
 	}
 	log.Infof("vCenter user: %s", admin.User)
 	log.Infof("PSC instance: %s", pscInstance)
-	log.Infof("PSC Admin domain: %s", pscDomain)
+	log.Infof("PSC domain: %s", pscDomain)
 
 	// Obtain the OVA VM's IP
 	vmIP, err := ip.FirstIPv4(ip.Eth0Interface)

